File: MATLAB/chemometrics/preprocessing.m
function X_processed = preprocessing(X, method, varargin)
%PREPROCESSING Aplica pré-processamento aos dados espectrais
%
% Sintaxe:
%   X_processed = preprocessing(X, method)
%   X_processed = preprocessing(X, method, 'Name', Value)
%
% Parâmetros:
%   X - Dados espectrais (amostras x variáveis)
%   method - Método de pré-processamento:
%           'snv' - Standard Normal Variate
%           'savgol' - Savitzky-Golay
%           'msc' - Multiplicative Scatter Correction
%           'mean_center' - Mean Centering
%           'autoscale' - Autoscaling
%
% Propriedades (Name-Value pairs):
%   'plot' - true/false para plotar resultados (default: false)
%   'window_length' - comprimento da janela para Savitzky-Golay (default: 15)
%   'polyorder' - ordem do polinômio para Savitzky-Golay (default: 2)
%   'deriv' - ordem da derivada para Savitzky-Golay (default: 0)
%   'reference' - espectro de referência para MSC (default: média)
%
% Exemplo:
%   X_processed = preprocessing(X, 'snv', 'plot', true);
%   X_processed = preprocessing(X, 'savgol', 'window_length', 21, 'deriv', 1);

    % Configurar parser de argumentos
    p = inputParser;
    addRequired(p, 'X', @isnumeric);
    addRequired(p, 'method', @ischar);
    addParameter(p, 'plot', false, @islogical);
    addParameter(p, 'window_length', 15, @isnumeric);
    addParameter(p, 'polyorder', 2, @isnumeric);
    addParameter(p, 'deriv', 0, @isnumeric);
    addParameter(p, 'reference', [], @isnumeric);
    
    parse(p, X, method, varargin{:});
    
    % Validar entrada
    if isempty(X)
        error('Dados de entrada não podem estar vazios.');
    end
    
    if size(X, 1) == 1
        X = X'; % Transpor se necessário
    end
    
    % Aplicar método de pré-processamento
    switch lower(method)
        case 'snv'
            X_processed = apply_snv(X);
        case 'savgol'
            X_processed = apply_savgol(X, p.Results.window_length, ...
                                     p.Results.polyorder, p.Results.deriv);
        case 'msc'
            X_processed = apply_msc(X, p.Results.reference);
        case 'mean_center'
            X_processed = apply_mean_center(X);
        case 'autoscale'
            X_processed = apply_autoscale(X);
        otherwise
            error('Método de pré-processamento não reconhecido: %s', method);
    end
    
    % Plotar resultados se solicitado
    if p.Results.plot
        plot_preprocessing_results(X, X_processed, method);
    end
end

function X_snv = apply_snv(X)
% Aplica Standard Normal Variate (SNV)
    % Calcular média e desvio padrão de cada espectro
    mean_spectrum = mean(X, 2);
    std_spectrum = std(X, 0, 2);
    
    % Evitar divisão por zero
    std_spectrum(std_spectrum == 0) = 1;
    
    % Aplicar SNV
    X_snv = (X - mean_spectrum) ./ std_spectrum;
end

function X_savgol = apply_savgol(X, window_length, polyorder, deriv)
% Aplica filtro Savitzky-Golay
    try
        X_savgol = zeros(size(X));
        for i = 1:size(X, 1)
            X_savgol(i, :) = sgolayfilt(X(i, :), polyorder, window_length, deriv);
        end
    catch ME
        warning('Erro no filtro Savitzky-Golay: %s', ME.message);
        X_savgol = X;
    end
end

function X_msc = apply_msc(X, reference)
% Aplica Multiplicative Scatter Correction (MSC)
    if isempty(reference)
        reference_spectrum = mean(X, 1);
    else
        reference_spectrum = reference;
    end
    
    % Centralizar os dados
    X_centered = X - mean(X, 2);
    reference_centered = reference_spectrum - mean(reference_spectrum);
    
    % Calcular coeficientes de regressão
    try
        coeffs = reference_centered' \ X_centered';
        
        % Ajustar os dados
        X_msc = (X_centered - coeffs' * reference_centered) + mean(X, 2);
    catch
        warning('Erro na decomposição QR. Retornando dados originais.');
        X_msc = X;
    end
end

function X_centered = apply_mean_center(X)
% Aplica centramento na média
    X_centered = X - mean(X, 1);
end

function X_scaled = apply_autoscale(X)
% Aplica autoscaling
    X_scaled = (X - mean(X, 1)) ./ std(X, 0, 1);
    
    % Evitar divisão por zero
    X_scaled(isnan(X_scaled)) = 0;
    X_scaled(isinf(X_scaled)) = 0;
end

function plot_preprocessing_results(X_original, X_processed, method)
% Plota os resultados do pré-processamento
    figure('Position', [100, 100, 1200, 600]);
    
    % Espectro original
    subplot(2, 1, 1);
    plot(X_original(1, :)', 'k-', 'LineWidth', 1);
    title('Espectro Original');
    ylabel('Absorbância');
    grid on;
    
    % Espectro processado
    subplot(2, 1, 2);
    plot(X_processed(1, :)', 'g-', 'LineWidth', 1);
    title(sprintf('Espectro Processado - %s', upper(method)));
    xlabel('Comprimento de Onda');
    ylabel('Absorbância');
    grid on;
    
    sgtitle(sprintf('Pré-processamento: %s', upper(method)));
end

File: Python/v2/chemometrics/preprocessing.py
"""
Módulo de pré-processamento espectroscópico. Mensagem 1
"""

import numpy as np
import matplotlib.pyplot as plt
from scipy.signal import savgol_filter
from scipy.sparse import eye, diags, csc_matrix
from scipy.sparse.linalg import spsolve
from typing import Optional, Tuple
import logging

logger = logging.getLogger(__name__)


class BasePreprocessor:
    """Classe base para pré-processadores."""
    
    def __init__(self, plot: bool = False):
        self.plot = plot
        self.is_fitted = False
        
    def _validate_input(self, X: np.ndarray) -> np.ndarray:
        """Valida e prepara os dados de entrada."""
        if not isinstance(X, np.ndarray):
            X = np.array(X, dtype=np.float64)
        
        if X.ndim == 1:
            X = X.reshape(1, -1)
            
        if X.size == 0:
            raise ValueError("Dados de entrada não podem estar vazios.")
            
        return X
    
    def _plot_results(self, X_original: np.ndarray, X_processed: np.ndarray, 
                     title: str, sample_idx: int = 1) -> None:
        """Plota os resultados do pré-processamento."""
        if not self.plot:
            return
            
        try:
            fig, (ax1, ax2) = plt.subplots(2, 1, figsize=(12, 6), dpi=300)
            
            # Espectro original
            ax1.plot(X_original[sample_idx, :].T, color='black', linewidth=1)
            ax1.set_title('Espectro Original')
            ax1.set_ylabel('Absorbância')
            ax1.grid(True, alpha=0.3)
            
            # Espectro processado
            ax2.plot(X_processed[sample_idx, :].T, color='green', linewidth=1)
            ax2.set_title(f'Espectro Processado - {title}')
            ax2.set_xlabel('Comprimento de Onda')
            ax2.set_ylabel('Absorbância')
            ax2.grid(True, alpha=0.3)
            
            plt.tight_layout()
            plt.show()
            
        except Exception as e:
            logger.warning(f"Erro ao plotar resultados: {e}")


class SNVPreprocessor(BasePreprocessor):
    """
    Standard Normal Variate (SNV) preprocessing.
    
    Normaliza cada espectro subtraindo a média e dividindo pelo desvio padrão.
    """
    
    def __init__(self, ddof: int = 1, plot: bool = False):
        super().__init__(plot)
        self.ddof = ddof
        
    def fit(self, X: np.ndarray) -> 'SNVPreprocessor':
        """Ajusta o pré-processador (SNV não precisa de ajuste)."""
        self._validate_input(X)
        self.is_fitted = True
        return self
    
    def transform(self, X: np.ndarray) -> np.ndarray:
        """Aplica SNV aos dados."""
        if not self.is_fitted:
            self.fit(X)
            
        X = self._validate_input(X)
        
        # Calcula média e desvio padrão de cada linha (espectro)
        mean_spectrum = np.mean(X, axis=1, keepdims=True)
        std_spectrum = np.std(X, axis=1, keepdims=True, ddof=self.ddof)
        
        # Evita divisão por zero
        std_spectrum = np.where(std_spectrum == 0, 1, std_spectrum)
        
        # Aplica SNV
        X_snv = (X - mean_spectrum) / std_spectrum
        
        if self.plot:
            self._plot_results(X, X_snv, 'SNV')
            
        return X_snv
    
    def fit_transform(self, X: np.ndarray) -> np.ndarray:
        """Ajusta e transforma os dados."""
        return self.fit(X).transform(X)


class SavitzkyGolayPreprocessor(BasePreprocessor):
    """
    Savitzky-Golay filter preprocessing.
    """
    
    def __init__(self, window_length: int = 15, polyorder: int = 2, 
                 deriv: int = 0, plot: bool = False):
        super().__init__(plot)
        self.window_length = window_length
        self.polyorder = polyorder
        self.deriv = deriv
        
    def fit(self, X: np.ndarray) -> 'SavitzkyGolayPreprocessor':
        """Ajusta o pré-processador."""
        self._validate_input(X)
        self.is_fitted = True
        return self
    
    def transform(self, X: np.ndarray) -> np.ndarray:
        """Aplica filtro Savitzky-Golay aos dados."""
        if not self.is_fitted:
            self.fit(X)
            
        X = self._validate_input(X)
        
        try:
            X_filtered = savgol_filter(X, self.window_length, self.polyorder, 
                                     deriv=self.deriv, axis=1)
        except Exception as e:
            logger.error(f"Erro no filtro Savitzky-Golay: {e}")
            return X
            
        if self.plot:
            title = f'Savitzky-Golay (window={self.window_length}, order={self.polyorder}, deriv={self.deriv})'
            self._plot_results(X, X_filtered, title)
            
        return X_filtered
    
    def fit_transform(self, X: np.ndarray) -> np.ndarray:
        """Ajusta e transforma os dados."""
        return self.fit(X).transform(X)


class MSCPreprocessor(BasePreprocessor):
    """
    Multiplicative Scatter Correction (MSC) preprocessing.
    """
    
    def __init__(self, reference: Optional[np.ndarray] = None, plot: bool = False):
        super().__init__(plot)
        self.reference = reference
        self.reference_spectrum = None
        
    def fit(self, X: np.ndarray) -> 'MSCPreprocessor':
        """Ajusta o pré-processador MSC."""
        X = self._validate_input(X)
        
        if self.reference is None:
            self.reference_spectrum = np.mean(X, axis=0)
        else:
            self.reference_spectrum = np.array(self.reference)
            
        self.is_fitted = True
        return self
    
    def transform(self, X: np.ndarray) -> np.ndarray:
        """Aplica MSC aos dados."""
        if not self.is_fitted:
            raise ValueError("Pré-processador não foi ajustado. Chame fit() primeiro.")
            
        X = self._validate_input(X)
        
        # Centralizar os dados
        X_centered = X - np.mean(X, axis=1, keepdims=True)
        reference_centered = self.reference_spectrum - np.mean(self.reference_spectrum)
        
        # Calcular coeficientes de regressão
        try:
            coeffs = np.linalg.lstsq(reference_centered[:, np.newaxis], X_centered.T, rcond=None)[0]
            
            # Ajustar os dados
            X_msc = (X_centered - coeffs.T * reference_centered) + np.mean(X, axis=1, keepdims=True)
            
        except np.linalg.LinAlgError:
            logger.warning("Erro na decomposição QR. Retornando dados originais.")
            X_msc = X
            
        if self.plot:
            self._plot_results(X, X_msc, 'MSC')
            
        return X_msc
    
    def fit_transform(self, X: np.ndarray) -> np.ndarray:
        """Ajusta e transforma os dados."""
        return self.fit(X).transform(X)


class MeanCenterPreprocessor(BasePreprocessor):
    """
    Mean Centering preprocessing.
    """
    
    def __init__(self, plot: bool = False):
        super().__init__(plot)
        self.mean_ = None
        
    def fit(self, X: np.ndarray) -> 'MeanCenterPreprocessor':
        """Ajusta o pré-processador."""
        X = self._validate_input(X)
        self.mean_ = np.mean(X, axis=0, keepdims=True)
        self.is_fitted = True
        return self
    
    def transform(self, X: np.ndarray) -> np.ndarray:
        """Aplica centramento na média aos dados."""
        if not self.is_fitted:
            raise ValueError("Pré-processador não foi ajustado. Chame fit() primeiro.")
            
        X = self._validate_input(X)
        X_centered = X - self.mean_
        
        if self.plot:
            self._plot_results(X, X_centered, 'Mean Centering')
            
        return X_centered
    
    def fit_transform(self, X: np.ndarray) -> np.ndarray:
        """Ajusta e transforma os dados."""
        return self.fit(X).transform(X)


class AutoScalePreprocessor(BasePreprocessor):
    """
    Autoscaling preprocessing.
    """
    
    def __init__(self, ddof: int = 1, plot: bool = False):
        super().__init__(plot)
        self.ddof = ddof
        self.mean_ = None
        self.std_ = None
        
    def fit(self, X: np.ndarray) -> 'AutoScalePreprocessor':
        """Ajusta o pré-processador."""
        X = self._validate_input(X)
        self.mean_ = np.mean(X, axis=0, keepdims=True)
        self.std_ = np.std(X, axis=0, keepdims=True, ddof=self.ddof)
        
        # Evita divisão por zero
        self.std_ = np.where(self.std_ == 0, 1, self.std_)
        
        self.is_fitted = True
        return self
    
    def transform(self, X: np.ndarray) -> np.ndarray:
        """Aplica autoscaling aos dados."""
        if not self.is_fitted:
            raise ValueError("Pré-processador não foi ajustado. Chame fit() primeiro.")
            
        X = self._validate_input(X)
        X_scaled = (X - self.mean_) / self.std_
        
        if self.plot:
            self._plot_results(X, X_scaled, 'Autoscaling')
            
        return X_scaled
    
    def fit_transform(self, X: np.ndarray) -> np.ndarray:
        """Ajusta e transforma os dados."""
        return self.fit(X).transform(X)
